src/crimson/os/seastore/seastore.cc

```diff
@@ -716,9 +722,15 @@ SeaStore::Shard::list_objects(CollectionRef ch,
 		  std::get<1>(ret) = std::get<1>(_ret);
 		  assert(limit >= next_objects.size());
 		  limit -= next_objects.size();
+		  LOG_PREFIX(SeaStore::list_objects);
+		  DEBUGT("got {} objects, left limit {}",
+		    t, next_objects.size(), limit);
+		  if (last && std::get<1>(ret) == pend) {
```

Comment:
Looks this is not consistent to https://github.com/ceph/ceph/pull/53062/files#r1338333338.

Assuming that `pend` is not `ghobject_t::get_max()` as an example, `OnodeManager::list_onodes()` will return `std::get<1>(ret) == pend` (this is case 3.) if and only if:
* The listing hasn't reached the very last of the tree (pend is not max).
* It has listed all the items in the range of `pstart` and `pend`.

Which means that `SeaStore::list_objects()` in this case also hasn't reached the very last of the tree, but it returns `ghobject_t::get_max()` below.

Reply:
I think I get your point, you mean that `OnodeManager::list_onodes()` and `SeaStore::list_objects()` should have the same semantics, but `OnodeManager::list_onodes()` returning `ghobject_t::max()` doesn't necessarily mean `SeaStore::list_objects()` should also return `ghobject_t::max()`, right?

Reply:
Done

Reply:
Correct, identical semantics to make sure they are sufficient to answer all the necessary questions.

But I assume the users (e.g. backfill scanning) might need to be adjusted if the original semantic is changed.

The actual listing implementations have to be different because `SeaStore::list_objects()` needs to handle two ranges internally.

Reply:
> The actual listing implementations have to be different because `SeaStore::list_objects()` needs to handle two ranges internally.

Yeah, I think this is done in the newly pushed code, please take a look.